pkg/util/encoding/binary.go

```diff
@@ -0,0 +1,43 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+// Package encoding is for utilities relating to the encoding package from the stdlib
+package encoding
+
+import (
+	"encoding"
+)
+
+// BinaryUnmarshalCallback returns a function that will decode the argument byte slice into *T
+// using `newFn` to create an instance of *T and the encoding.BinaryUnmarshaler interface to do the actual conversion.
+// `callback` will be called with the resulting *T.
+// If the argument byte slice is empty, callback will be called with `nil`.
+// Unmarshalling errors will be provided to the callback as the second argument. The data argument to the callback
+// may still be non-nil even if there was an error. This allows the callback to handle the allocated object, even
+// in the face of errors.
+// This function panics if `*T` does not implement encoding.BinaryUnmarshaler.
+func BinaryUnmarshalCallback[T any](newFn func() *T, callback func(*T, error)) func(buf []byte) {
```

Comment:
Sharing some thoughts here so fell free to ignore this comment.

I agree that having panic is not necessary an issue. `panic` could be used for example to reflect a programming error (typically when a contract is not respected like in your case). I think `panic` should be restricted to non conditional code path, ie, when running the application and it `panic` immediately, if the contract is not respected. If it panics only if a config option is enabled or understand specific condition then `panic` may not be the most appropriate solution.

Did you consider the following implementation which also allows `nil` for the callback?
```go
func BinaryUnmarshalCallback[T encoding.BinaryUnmarshaler](newFn func() T, callback func(T, error)) func(buf []byte) {
	var v T
	callback(v, nil)
	return nil
}
```

Reply:
One thing I'm trying to guard against is someone trying to implement `encoding.BinaryUnmarshaler` on a non-pointer receiver, which is basically guaranteed to not work correctly. I cannot seem to find way with generics to enforce both that the interface is implemented AND it is implemented on a pointer receiver. See https://go.dev/play/p/HRqQPb6xvDJ for a subtle way in which it can fail.

Reply:
I have given up and removed the panic. I'm just going to assume they have implemented `encoding.BinaryUnmarshaler` correctly because it seems Go is automatically calling the value receiver function, even if given a pointer.